Conversion to a different nextToken implementation

squeek502 opened the pull request:
This is unfinished, but I'm submitting it now since it's a big change and I don't want to get ahead of myself if it's not something you think is worthwhile.

---

This looks like more changes than there actually are, since the addition of a while loop changed the indentation level within nextToken.

The basic ideas:

---

- ~~Removal of Token.EndOfFile. Instead, nextToken returning `null` is *always* EOF~~
- `nextToken` will now *never* return null
  + nextToken will loop until it has either an error or a token ready for consumption instead of returning intermediate null tokens
  + Whenever we know a token or error exists, we return it from the nextToken function as soon as we're done setting up the tokenizer state for the next character/token. This either means popping from the queue or returning a single error/token directly.

Examples:
```zig
'>' => {
    self.addAttributeToCurrentToken();
    self.state = .Data;
    self.emitToken(self.currentToken.?);
    self.currentToken = null;
    return self.popQueuedErrorOrToken(); // this line is new
},
```

(returning a single error is the same as before)
```zig
0x00 => {
    self.emitToken(Token { .Character = .{ .data = 0x00  } });
    return ParseError.UnexpectedNullCharacter;
},
```

---

- Addition of Tokenizer.errorQueue that can be used to return multiple errors at once (closes https://github.com/watzon/zhtml/issues/3)

Example:
```zig
    pub fn emitToken(self: *Self, token: Token) void {
        if (token == .EndTag) {
            if (token.EndTag.attributes.items().len > 0) {
                self.errorQueue.writeItem(@errorToInt(ParseError.EndTagWithAttributes)) catch unreachable;
            }
            if (token.EndTag.selfClosing) {
                self.errorQueue.writeItem(@errorToInt(ParseError.EndTagWithTrailingSolidus)) catch unreachable;
            }
        }
        self.backlog.writeItem(token) catch unreachable;
    }
```

---

Benefits:
- Allows for more straightforward handling on the caller's side of nextToken
  + Makes it more similar to other `next` functions in Zig that return null once they are finished iterating
- Allows for implementing something like "When an end tag token is emitted with attributes, that is an end-tag-with-attributes parse error." from the spec without a lot of special-casing, since we can just append the errors in emitToken and know that we're going to immediately return the errors along with the token afterwards

Drawbacks:
- Potentially leads to a more complicated to follow nextToken implementation since it's a loop rather than a single switch per call

---

This does still successfully parse the old `test.html` that was in the repo, and it passes a few more html5lib tests that were failing.

Things left to do:
- [x] Integrate with `flushTemporaryBufferAsCharacterReference`/`flushCodepointAsCharacterReference` and related states
- [x] General cleanup/removal of debug printing
- [x] Make sure infinite loops are impossible (this also relates to EOF handling which I've left unchanged for now)
  + Any infinite loops should be bugs in the implementation (usually jumping to the wrong state for a given case)

squeek502 committed:
Start of conversion to a different nextToken implementation

This looks like more changes than there actually are, since the addition of a while loop changed the indentation level within nextToken.

The basic ideas:
- Removal of Token.EndOfFile, instead nextToken returning null is *always* EOF
  + nextToken will loop until it has either an error or a token ready for consumption instead of returning intermediate null tokens
  + Whenever we know a token or error exists, we return it from the nextToken function as soon as we're done setting up the tokenizer state for the next character/token
- Addition of Tokenizer.errorQueue that can be used to return multiple errors at once (https://github.com/watzon/zhtml/issues/3)

Benefits:
- Allows for more straightforward handling on the caller's side of nextToken, makes it more similar to other next functions in Zig that return null once they are finished iterating
- Allows for implementing something like "When an end tag token is emitted with attributes, that is an end-tag-with-attributes parse error." from the spec without a lot of special-casing, since we can just append the errors in emitToken and know that we're going to immediately return the errors afterwards

Drawbacks:
- Potentially leads to a more complicated to follow nextToken implementation since it's a loop rather than a single switch per call

watzon commented:
I like this, it solves a lot of issues! Go ahead with it. I was working on my own thing last night, but I like this better.

squeek502 committed:
Don't compare EndTag attributes in expectEqualTokens

EndTag attributes are ultimately ignored; we only store them to be able to emit EndTagWithAttributes errors

squeek502 committed:
Integrate character references with popQueuedErrorOrToken paradigm

squeek502 committed:
Move eof check inside the loop

squeek502 committed:
Make netChar/peekChar/etc handle EOF and add tests

squeek502 committed:
Reintroduce Token.EndOfFile, make nextToken never return null

Didn't realize the spec actually says 'emit an end-of-file token', so having an actual EndOfFile token makes more sense to me than returning null as an EndOfFile surrogate. Having it be an actual token also allows for adding EndOfFile to the backlog if that is every necessary somewhere.

squeek502 committed:
Example of handling EOF in a state

All the other states will need similar handling (for every call of nextChar/peekChar/peekN). This is just an example

squeek502 commented:
Did some more work. Note: this branch will not currently compile. Here are the important bits from the commit messages:

>  [Reintroduce Token.EndOfFile, make nextToken never return null](https://github.com/watzon/zhtml/pull/6/commits/d741fe8d829787381b1740a41bbec7dadbe7d53f)

> Didn't realize the spec actually says 'emit an end-of-file token', so having an actual EndOfFile token makes more sense to me than returning null as an EndOfFile surrogate. Having it be an actual token also allows for adding EndOfFile to the backlog if that is ever necessary somewhere.

---

> [Make nextChar/peekChar/etc handle EOF and add tests](https://github.com/watzon/zhtml/pull/6/commits/23d846b53caeb216491bd82b2f820a0402de6e90)

This means that nextChar/peekChar/etc now return optional values, where `null` means that EOF was reached. This means that every caller of these functions (i.e. every state) needs to handle the possibility of `null` and react accordingly. I've [updated the data state as an example](https://github.com/watzon/zhtml/pull/6/commits/b21ff6d95ee56813c6fc98a46c55b1b8faaaf481), but all states will need to be updated similarly.

Once that's done (a rather large task) then this should be mergable. I won't be able to finish this part up until the weekend, so feel free to commit to this branch if you'd like to help with it (you should be able to push to this branch of my fork directly since I enabled that when submitting the PR).

---

Side note about EOF handling: Another possible way to handle EOF would be to change the loop in `nextToken` to `while(!self.eof())` and then add an `else` case to the loop (which only gets evaluated if the while condition is the reason for the loop exiting [in other words, it didn't exit due to a `break;`]) that then has its own switch to handle all the EOF cases. This is [what the Zig tokenizer does](https://github.com/ziglang/zig/blob/2064e84cdd596b5bc63c8d3e7e5aa4130221faee/lib/std/zig/tokenizer.zig#L1288-L1404).

This won't work for this parser, however, because it's possible to need to handle EOF by reconsuming, so we'd need to basically wrap the loop in another loop or some such craziness which doesn't feel worth it anymore. Plus, handling EOF per-state at the same time as all the other possible characters is more in line with how the spec is structured anyway.

watzon commented:
Good work, and thanks for your help with this!

squeek502 committed:
Update the rest of the switch to handle EOF

This just gets it to compile again, things are still quite broken

squeek502 committed:
Rework next/peek/current again to actually be right this time

My last attempt at this was not right at all. It didn't handle the beginning/end anywhere close to properly

squeek502 committed:
Add missing '>' case to ScriptDataEscapedDashDash state

squeek502 committed:
A few fixes

squeek502 committed:
Fix peekN when reconsuming and fix test

squeek502 committed:
Use hasQueuedErrorOrToken instead of bool return

squeek502 commented:
Ok, I think this is in a good enough state to merge now. There are still bugs for sure that need to be worked out, but the general idea is implemented.

Some notes:
- Character reference states have no EOF handling. I left all EOF handling as `// TODO` in those cases for now since how characters are consumed will need to change in the future in order to be spec compliant (see [this added comment](https://github.com/watzon/zhtml/blob/d9ee37b3a9a6c9c3983f7d070264359636aed3aa/src/tokenizer.zig#L1947-L1948))
- It seems like there are bugs in how/when token data is created and then added to tokens (i.e. sometimes the code will forget to set some field of a token before emitting it). It might be nice to attempt to move the temporary buffers onto the tokens themselves or in some other way centralize how things like `commentData` are related to tokens to avoid these types of bugs/duplicate code.
  + EDIT: I will have a follow-up PR for this
- When debugging, adding `std.debug.print("State: {} peekChar: {c}\n", .{self.state, self.peekChar()});` as the first line within the `while(true)` loop in `nextToken` is helpful.
- `column` field incrementing still needs some work, especially for EOF related errors

squeek502 committed:
Remove useless std.math.min call

squeek502 committed:
Fix peekChar for second-to-last character
